core/services/ocr2/plugins/promwrapper/plugin.go

```diff
@@ -80,93 +96,251 @@ var (
+	promQuery.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPrometheusBackend) SetObservationDuration(labelValues []string, duration float64) {
+	promObservation.WithLabelValues(labelValues...).Observe(duration)
 }
 
-func New(plugin types.ReportingPlugin, name string, chainType string, chainID *big.Int, config types.ReportingPluginConfig) types.ReportingPlugin {
+func (*defaultPrometheusBackend) SetReportDuration(labelValues []string, duration float64) {
+	promReport.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPrometheusBackend) SetShouldAcceptFinalizedReportDuration(labelValues []string, duration float64) {
+	promShouldAcceptFinalizedReport.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPrometheusBackend) SetShouldTransmitAcceptedReportDuration(labelValues []string, duration float64) {
+	promShouldTransmitAcceptedReport.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPrometheusBackend) SetCloseDuration(labelValues []string, duration float64) {
+	promClose.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPrometheusBackend) SetQueryToObservationLatency(labelValues []string, latency float64) {
+	promQueryToObservationLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPrometheusBackend) SetObservationToReportLatency(labelValues []string, latency float64) {
+	promObservationToReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPrometheusBackend) SetReportToAcceptFinalizedReportLatency(labelValues []string, latency float64) {
+	promReportToAcceptFinalizedReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPrometheusBackend) SetAcceptFinalizedReportToTransmitAcceptedReportLatency(labelValues []string, latency float64) {
+	promAcceptFinalizedReportToTransmitAcceptedReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func New(
+	plugin types.ReportingPlugin,
+	name string,
+	chainType string,
+	chainID *big.Int,
+	config types.ReportingPluginConfig,
+	backend PrometheusBackend,
+) types.ReportingPlugin {
+	// Apply passed-in Prometheus backend if one is given.
+	var prometheusBackend PrometheusBackend = &defaultPrometheusBackend{}
+	if backend != nil {
+		prometheusBackend = backend
+	}
+
 	return &promPlugin{
-		wrapped:      plugin,
-		name:         name,
-		chainType:    chainType,
-		chainID:      chainID,
-		oracleID:     fmt.Sprintf("%d", config.OracleID),
-		configDigest: common.Bytes2Hex(config.ConfigDigest[:]),
+		wrapped:           plugin,
+		name:              name,
+		chainType:         chainType,
+		chainID:           chainID,
+		oracleID:          fmt.Sprintf("%d", config.OracleID),
+		configDigest:      common.Bytes2Hex(config.ConfigDigest[:]),
+		prometheusBackend: prometheusBackend,
 	}
 }
 
 func (p *promPlugin) Query(ctx context.Context, timestamp types.ReportTimestamp) (types.Query, error) {
 	start := time.Now().UTC()
```

Comment:
Make me wonder - should we have a timer for when one round ends and another begins (I think it would be measuring from `Report` to the next `Query` call)? Would that be useful? Or is that just `deltaRound`?

Reply:
We could do such a thing, though I think that may be out of scope for this PR. Off-the-top of my head, it would involve a mapping of the report timestamp, and then looking up the report timestamp with the same epoch and the round just before it, or the first round of an epoch and then the last round for an epoch. Could be done, but also maybe this is already predetermined by OCR..